prestoadmin/util/version_util.py

```diff
@@ -0,0 +1,189 @@
+
+    return tuple(result)
+
+
+class VersionRange(object):
+    """
+    Represents a range of version numbers [min_version, max_version).
+    The interval is right-open so that you can construct a numerically
+    continuous list of versions like so:
+    l = [VersionRange((0, 0), (0, 5)),
+         VersionRange((0, 5), (1, 0))]
+    and for all versions v where 0.0 <= v < 1.0 is contained in exactly one
+    VersionRange in l.
+
+    Continuity between version ranges can be checked using is_continuous.
+
+    VersionRanges understand how to check if a Teradata version is contained
+    in a VersionRange, but do no special handling to accomodate Teradata
+    versions in their internal min_version and max_version members. I.e.,
+    creating a VersionRange with a Teradata version will work, but __contains__
+    will not work correctly. We don't currently need this, and hope not to.
+
+    Note that the right-open interval representation of a version range does
+    NOT allow the creation of a VersionRange that contains exactly one version.
+
+    Note that empty intervals cannot be constructed as the serve no useful
+    purpose. Specifically, we assert that min_version < max_version in the
+    constructor.
+    """
+
+    def __init__(self, min_version, max_version, versioned_thing=None):
+        # not pythonic, but bare ints screw things up.
+        assert isinstance(min_version, tuple)
+        assert isinstance(max_version, tuple)
+        l = max(len(min_version), len(max_version))
+        min_pad = VersionRange.pad_tuple(min_version, l, 0)
+        max_pad = VersionRange.pad_tuple(max_version, l, 0)
+        assert min_pad < max_pad
+        self.min_version = min_version
+        self.max_version = max_version
+        self.versioned_thing = versioned_thing
+
+    def __str__(self):
+        return '[%s, %s) -> %s' % (
+            '.'.join([str(c) for c in self.min_version]),
+            '.'.join([str(c) for c in self.max_version]),
+            self.versioned_thing)
+
+    @staticmethod
+    def strip_td_suffix(other):
+        last_component = other[-1]
+        if isinstance(last_component, int):
+            return other
+
+        if TD_VERSION.match(last_component):
+            new_last = last_component[:-1]
+            return other[:-1] + (int(new_last),)
+
+        return other
+
+    @staticmethod
+    def pad_tuple(tup, length, pad):
+        assert len(tup) <= length
+        result = list(tup)
+        while len(result) < length:
+            result.append(pad)
+        return tuple(result)
+
+    def zero_pad(self, other):
+        """
+        Pad out min_version, max_version, and other with zeroes to the length
+        of the longest of the three. This allows subsequent comparisons to work
+        as expected when tuples are of unequal length.
+        Returns a tuple of tuples padded out to the same length
+        """
+        l = max(len(self.min_version), len(self.max_version), len(other))
```

Comment:
It's needed for correct comparisons when the tuples are the same modulo trailing zeroes:
>>> (1,) == (1, 0, 0)
False
>>> (1,) < (1, 0, 0)
True
>>> (1,) > (1, 0, 0)
False